pkgs/build-support/neowrap.nix

```diff
@@ -0,0 +1,294 @@
+      linkPaths = {
+        XDG_DATA_DIRS = "$out/share";
+        GI_TYPELIB_PATH = "$out/lib/girepository-1.0";
+        QT_PLUGIN_PATH = "$out/${pkgs.qt5.qtbase.qtPluginPrefix}";
+        QML2_IMPORT_PATH = "$out/${pkgs.qt5.qtbase.qtQmlPrefix}";
+        # Not using a strict version here as it may confuse someone digging
+        # into the results of this wrapper
+        PYTHONPATH = "$out/python-sitePackages";
+      };
+      # If you want an environment variable to have a single value and that's it,
+      # put it here:
+      singleValue = [
+        "GDK_PIXBUF_MODULE_FILE"
+      ];
+      # To prevent a stack overflow when inspecting a derivation's inputs
+      # recursively, we use this list to tell the function that dives into the
+      # dependency graph, what packages should be skipped instantly.
+      skipDivingInto = with pkgs; [
+        gnu-efi iptables kexectools libapparmor libcap libidn2 libmicrohttpd
+        libnetfilter_conntrack libnftnl libseccomp pciutils systemd acl autogen
+        gnutls guile libevent p11-kit unbound xorg.libXext freetype xorg.libICE
+        libpng_apng kmod libgcrypt curl libkrb5 libssh2 nghttp2 libtool nettle
+        glib libselinux utillinux xorg.libX11 xorg.libxcb xorg.libXdmcp libxslt cracklib
+        linux-pam xorg.libXau libxml2 sqlite readline bison zlib libffi openssl ncurses
+        # Note we add perl itself here, not any library of it
+        perl xorg.xorgproto expat xz fontconfig xorg.libXfixes llvm
+      ];
+    };
+    # recursive function that goes deep through the dependency graph of a given
+    # list of packages and creates a list of all buildInputs they all depend
+    # on. The 2nd argument pkgsFound is used internally and it's expected to be
+    # [] at the first call.
+    getAllInputs =
+      pkgs:
+      pkgsFound:
+      map (
+        pkg:
+        if builtins.isAttrs pkg then
+          let
+            inEncyclopedia =
+              inputPkg:
+              lib.lists.any (
+                knownPkg:
+                if builtins.isAttrs inputPkg then
+                  # builtins.trace "testing ${inputPkg.name} if in encyclopedia: ${builtins.toJSON (inputPkg.name == knownPkg.name)}" inputPkg.name == knownPkg.name
+                  inputPkg.name == knownPkg.name
+                else
+                  false
+                ) encyclopedia.skipDivingInto
+            ;
+            notInEncyclopedia = 
+              inputPkg:
+              !(inEncyclopedia inputPkg)
+            ;
+            hasInputs = pkg.buildInputs != [] || pkg.propagatedBuildInputs != [];
+          in
+            if !(inEncyclopedia pkg) && hasInputs then
+              let
+                # From some reason, pkg == knownPkg doesn't work :/ don't know
+                # why... Never the less this should be good enough
+                deeperPkgs = builtins.filter notInEncyclopedia (pkg.buildInputs ++ pkg.propagatedBuildInputs);
```

Comment:
How does the neowrapper handle propagation boundaries when `buildInputs` and `propagatedBuildInputs` are the same?
For example, `libinput` depends on `pyudev` for its utilities but we do not want everything that links against `libinput` to have `pyudev` as dependency.

Reply:
`libinput` not only has a certain python environment in it's `buildInputs`, it also references this python env at `${libinput.bin}/libexec/libinput/libinput-measure-fuzz`. Note it's the `bin` output, not the more widely linked against `out` output.

> we do not want everything that links against libinput to have pyudev as dependency.

Agreed. The current design suffers from the inability to distinguish between `libinput` specified in a drv's `buildInputs`, vs `libinput.out` (and not `libinput.bin`) referenced by a drv. With the ability to distinguish between the two, `wrapGenric` should be capable of knowing not to add python related env vars to a wrapper, if it's only `libinput.out` that's referenced.

This issue naturally relates to the [unresolved questions of the RFC](https://github.com/doronbehar/rfcs/blob/declarative-wrappers/rfcs/0075-declarative-wrappers.md#unresolved-questions).